src/com/amazon/ion/BufferConfiguration.java

```diff
@@ -0,0 +1,186 @@
+package com.amazon.ion;
+
+/**
+ * Provides logic common to all BufferConfiguration implementations.
+ * @param <T> the {@link BufferEventHandler} implementation applicable to the concrete
+ *            ReaderLookaheadBufferBase subclass.
+ * @param <U> the type of the concrete subclass of this BufferConfiguration that is applicable to the
+ *            ReaderLookaheadBufferBase subclass.
+ */
+public abstract class BufferConfiguration<T extends BufferEventHandler, U extends BufferConfiguration<T, U>> {
+
+    /**
+     * Provides logic common to all BufferConfiguration Builder implementations.
+     * @param <T> the type of {@link BufferEventHandler} used by the BufferConfiguration.
+     * @param <U> the type of BufferConfiguration.
+     * @param <V> the type of Builder that builds BufferConfiguration subclasses of type T.
+     */
+    public static abstract class Builder<
+        T extends BufferEventHandler,
+        U extends BufferConfiguration<T, U>,
+        V extends Builder<T, U, V>
+    > {
```

Comment:
The reason why I like the single letters for type parameters is because it makes it very clear to me when I'm dealing with a parameter vs a concrete type (I would never give a class or interface a single letter name). However, I will concede that this has drawbacks, e.g. requiring a mental mapping, hurting readability for new readers of the code.
 
I've now noticed that IntelliJ (and presumably other IDEs) highlight type parameters with a different color than concrete types, which nullifies my initial point. I've gone ahead and made the suggested change.